File: subprojects/robotpy-commands-v2/commands2/src/cpp/frc2/command/button/CommandGenericHID.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc2/command/button/CommandGenericHID.h"

using namespace frc2;

#define loop (loop_arg ? *loop_arg : CommandScheduler::GetInstance().GetDefaultButtonLoop())

Trigger CommandGenericHID::Button(int button, std::optional<frc::EventLoop*> loop_arg) const {
  return GenericHID::Button(button, loop).CastTo<Trigger>();
}

Trigger CommandGenericHID::POV(int angle, std::optional<frc::EventLoop*> loop_arg) const {
  return POV(0, angle, loop);
}

Trigger CommandGenericHID::POV(int pov, int angle, std::optional<frc::EventLoop*> loop_arg) const {
  return Trigger(loop,
                 [this, pov, angle] { return this->GetPOV(pov) == angle; });
}

Trigger CommandGenericHID::POVUp(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(0, loop);
}

Trigger CommandGenericHID::POVUpRight(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(45, loop);
}

Trigger CommandGenericHID::POVRight(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(90, loop);
}

Trigger CommandGenericHID::POVDownRight(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(135, loop);
}

Trigger CommandGenericHID::POVDown(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(180, loop);
}

Trigger CommandGenericHID::POVDownLeft(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(225, loop);
}

Trigger CommandGenericHID::POVLeft(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(270, loop);
}

Trigger CommandGenericHID::POVUpLeft(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(315, loop);
}

Trigger CommandGenericHID::POVCenter(std::optional<frc::EventLoop*> loop_arg) const {
  return POV(360, loop);
}

Trigger CommandGenericHID::AxisLessThan(int axis, double threshold,
                                        std::optional<frc::EventLoop*> loop_arg) const {
  return Trigger(loop, [this, axis, threshold]() {
    return this->GetRawAxis(axis) < threshold;
  });
}

Trigger CommandGenericHID::AxisGreaterThan(int axis, double threshold,
                                           std::optional<frc::EventLoop*> loop_arg) const {
  return Trigger(loop, [this, axis, threshold]() {
    return this->GetRawAxis(axis) > threshold;
  });
}

File: subprojects/robotpy-commands-v2/commands2/button/commandgenerichid.py
from typing import Optional

from wpilib.event import EventLoop
from wpilib.interfaces import GenericHID

from ..commandscheduler import CommandScheduler
from .trigger import Trigger


class CommandGenericHID:
    """
    A version of GenericHID with Trigger factories for command-based.
    """

    def __init__(self, port: int):
        """
        Construct an instance of a device.

        :param port: The port on the Driver Station that the device is plugged into.
        """
        self._hid = GenericHID(port)

    def getHID(self) -> GenericHID:
        """
        Get the underlying GenericHID object.
        """
        return self._hid

    def button(self, button: int, loop: Optional[EventLoop] = None) -> Trigger:
        """
        Constructs an event instance around this button's digital signal.

        :param button: The button index
        :param loop: the event loop instance to attache the event to.
        """
        if loop is None:
            loop = CommandScheduler.getInstance().getDefaultButtonLoop()
        return Trigger(loop, lambda: self._hid.getRawButtonPressed(button))

    def pov(
        self, angle: int, *, pov: int = 0, loop: Optional[EventLoop] = None
    ) -> Trigger:
        """
        Constructs a Trigger instance based around this angle of a POV on the HID.

        The POV angles start at 0 in the up direction, and increase clockwise (e.g. right is 90,
        upper-left is 315).

        :param angle: POV angle in degrees, or -1 for the center / not pressed.
        :param pov: index of the POV to read (starting at 0). Defaults to 0.
        :param loop: the event loop instance to attach the event to. Defaults to {@link
            CommandScheduler#getDefaultButtonLoop() the default command scheduler button loop}.
        :returns: a Trigger instance based around this angle of a POV on the HID.
        """
        if loop is None:
            loop = CommandScheduler.getInstance().getDefaultButtonLoop()
        return Trigger(loop, lambda: self._hid.getPOV(pov) == angle)

    def povUp(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 0 degree angle (up) of the default (index 0) POV
        on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the default command
        scheduler button loop}.

        :returns: a Trigger instance based around the 0 degree angle of a POV on the HID.
        """
        return self.pov(0)

    def povUpRight(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 45 degree angle (right up) of the default (index
        0) POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the default
        command scheduler button loop}.

        :returns: a Trigger instance based around the 45 degree angle of a POV on the HID.
        """
        return self.pov(45)

    def povRight(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 90 degree angle (right) of the default (index 0)
        POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the default command
        scheduler button loop}.

        :returns: a Trigger instance based around the 90 degree angle of a POV on the HID.
        """
        return self.pov(90)

    def povDownRight(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 135 degree angle (right down) of the default
        (index 0) POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the
        default command scheduler button loop}.

        :returns: a Trigger instance based around the 135 degree angle of a POV on the HID.
        """
        return self.pov(135)

    def povDown(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 180 degree angle (down) of the default (index 0)
        POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the default command
        scheduler button loop}.

        :returns: a Trigger instance based around the 180 degree angle of a POV on the HID.
        """
        return self.pov(180)

    def povDownLeft(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 225 degree angle (down left) of the default
        (index 0) POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the
        default command scheduler button loop}.

        :returns: a Trigger instance based around the 225 degree angle of a POV on the HID.
        """
        return self.pov(225)

    def povLeft(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 270 degree angle (left) of the default (index 0)
        POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the default command
        scheduler button loop}.

        :returns: a Trigger instance based around the 270 degree angle of a POV on the HID.
        """
        return self.pov(270)

    def povUpLeft(self) -> Trigger:
        """
        Constructs a Trigger instance based around the 315 degree angle (left up) of the default (index
        0) POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the default
        command scheduler button loop}.

        :returns: a Trigger instance based around the 315 degree angle of a POV on the HID.
        """
        return self.pov(315)

    def povCenter(self) -> Trigger:
        """
        Constructs a Trigger instance based around the center (not pressed) position of the default
        (index 0) POV on the HID, attached to {@link CommandScheduler#getDefaultButtonLoop() the
        default command scheduler button loop}.

        :returns: a Trigger instance based around the center position of a POV on the HID.
        """
        return self.pov(-1)

    def axisLessThan(
        self, axis: int, threshold: float, loop: Optional[EventLoop] = None
    ) -> Trigger:
        """
        Constructs a Trigger instance that is true when the axis value is less than {@code threshold},
        attached to the given loop.

        :param axis: The axis to read, starting at 0
        :param threshold: The value below which this trigger should return true.
        :param loop: the event loop instance to attach the trigger to
        :returns: a Trigger instance that is true when the axis value is less than the provided
            threshold.
        """
        if loop is None:
            loop = CommandScheduler.getInstance().getDefaultButtonLoop()
        return Trigger(loop, lambda: self._hid.getRawAxis(axis) < threshold)

    def axisGreaterThan(
        self, axis: int, threshold: float, loop: Optional[EventLoop] = None
    ) -> Trigger:
        """
        Constructs a Trigger instance that is true when the axis value is greater than {@code
        threshold}, attached to the given loop.

        :param axis: The axis to read, starting at 0
        :param threshold: The value above which this trigger should return true.
        :param loop: the event loop instance to attach the trigger to.
        :returns: a Trigger instance that is true when the axis value is greater than the provided
            threshold.
        """
        if loop is None:
            loop = CommandScheduler.getInstance().getDefaultButtonLoop()
        return Trigger(loop, lambda: self._hid.getRawAxis(axis) > threshold)

    def getRawAxis(self, axis: int) -> float:
        """
        Get the value of the axis.

        :param axis: The axis to read, starting at 0.
        :returns: The value of the axis.
        """
        return self._hid.getRawAxis(axis)
